akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolGateway.scala

```diff
@@ -39,56 +41,74 @@ private object PoolGateway {
  * get reused will automatically forward requests directed at them to the latest pool incarnation from the cache.
  */
 private[http] class PoolGateway(hcps: HostConnectionPoolSetup,
-                                _shutdownStartedPromise: Promise[Done])( // constructor arg only
+                                _shutdownStartedPromise: Option[Promise[Done]] = None)( // used for cache invalidation
                                   implicit system: ActorSystem, fm: Materializer) {
   import PoolGateway._
   import fm.executionContext
 
-  private val state = {
+  private[this] var state: State = {
     val shutdownCompletedPromise = Promise[Done]()
     val props = Props(new PoolInterfaceActor(hcps, shutdownCompletedPromise, this)).withDeploy(Deploy.local)
     val ref = system.actorOf(props, PoolInterfaceActor.name.next())
-    new AtomicReference[State](Running(ref, _shutdownStartedPromise, shutdownCompletedPromise))
+    Running(ref, _shutdownStartedPromise, shutdownCompletedPromise)
   }
 
-  def currentState: Any = state.get() // enables test access
+  private[this] val lock = new ReentrantLock(true)
 
-  def apply(request: HttpRequest, previousIncarnation: PoolGateway = null): Future[HttpResponse] =
-    state.get match {
+  def currentState: Any = state // enables test access
+
+  private def setCurrentState(newState: State) = {
+    lock.lock()
+    state = newState
+    lock.unlock()
+  }
+
+  def apply(request: HttpRequest, previousIncarnation: PoolGateway = null): Future[HttpResponse] = {
+    lock.lock()
+    state match {
       case Running(ref, _, _) ⇒
         val responsePromise = Promise[HttpResponse]()
         ref ! PoolInterfaceActor.PoolRequest(request, responsePromise)
+        lock.unlock()
```

Comment:
unlock may not get called if Running-extractor throws, or if it cannot allocate a Promise of PoolRequest or if !-throws.